release/models/gnsi/openconfig-gnsi-credentialz.yang

```diff
@@ -0,0 +1,289 @@
+
+  grouping user-ssh-credentials-version {
+    description
+      "System role credentials freshness-related data.";
+
+    leaf authorized-principals-list-version {
+      type version;
+      description
+        "The version of the list of authorized principals currently
+        associated with this system role.";
+    }
+
+    leaf authorized-principals-list-created-on {
+      type created-on;
+      description
+        "The timestamp of the moment the currently used list of
+        authorized principals has been created.";
+    }
+
+    leaf authorized-keys-list-version {
+      type version;
+      description
+        "The version of the list of authorized keys that is currently
+        associated with this system role.";
+    }
+
+    leaf authorized-keys-list-created-on {
+      type created-on;
+      description
+        "The timestamp of the moment the currently used list of
+        authorized keys has been created.";
+    }
+  }
+
+  grouping console-config-state {
+    description
+      "Console-related configuration and state.";
+    container console {
+      description
+        "Console-related configuration and state.";
+
+      container config {
+        description
+          "Console-related configuration.";
+      }
+
+      container state {
+        config false;
+        description
+          "Console-related state.";
+
+        uses counters;
+      }
+    }
+  }
+  // System role console related definitions.
+
+  grouping user-console-credentials-version {
+    description
+      "System role credentials freshness-related data.";
+
+    leaf password-version {
+      type version;
+      description
+        "The version of the password that is currently used to
+        authenticate this user account.";
+    }
+
+    leaf password-created-on {
+      type created-on;
+      description
+        "The timestamp of the moment the currently used password has
+        been created.";
+    }
+  }
+
+  // Augments section.
+
+  augment "/oc-sys:system" {
+    description
+      "Console credentials freshness data.";
+
+    uses console-config-state;
+  }
+  augment "/oc-sys:system/oc-sys:ssh-server/oc-sys:state" {
+    description
+      "SSH server credentials freshness data.";
+
+    uses ssh-server-credentials-version;
+    uses counters;
+  }
+  augment "/oc-sys:system/oc-sys:aaa/oc-sys:authentication/oc-sys:users/" +
+          "oc-sys:user/oc-sys:state" {
+    description
+      "A system role credentials freshness information.";
+
+    uses user-console-credentials-version;
+    uses user-ssh-credentials-version;
+  }
+  augment "/oc-sys:system/oc-sys:aaa/oc-sys:authentication/oc-sys:admin-user/" +
+          "oc-sys:state" {
+    description
+      "A system role credentials freshness information.";
+
+    uses user-console-credentials-version;
+    uses user-ssh-credentials-version;
+  }
+  deviation "/oc-sys:system/oc-sys:aaa/oc-sys:authentication/oc-sys:users/" +
```

Comment:
These should be a when not deviate

Reply:
@dplore I think this is a fair proposal if we're going to continue keeping all models within the same repository. The benefit is it enables annotating conflicts between models using `when` rather than just comments, which is what we decided to do here with the WiFi models: https://github.com/openconfig/public/issues/882

However, there are some tooling implications which will delay this PR a bit:
- [ ] goyang currently unconditionally accepts `when` statements, but it should unconditionally *ignore* deviated nodes that are conditioned with `when`, since that's the opposite case. This will fix the breaking change CI.
- [ ] ygot depends on goyang, and we must ensure that it also ignores conditioned deviate statements in the generated code.

If this makes sense to you, I can work on this tooling change this or next week to enable this.

Reply:
I agree that some usage of `when` to indicate mutual-exclusion is much better than using deviate.   (We should not have deviations within the set of OpenConfig models)

Is `when gnsi is enabled`  deviate not-supported a valid approach to define mutual exclusion?  It seems odd to have a runtime conditional create schema changes?

Reply:
Sorry, I believe I misunderstood the suggestion.

The problem is gNSI wishes to disable some AAA leafs when it is enabled, but bare deviation statements prevent any system from implementing the union of all modules (see https://github.com/openconfig/public/issues/882).

My comment above is based on adding a `when` guard to the deviate statements, which solves the all-modules problem:
```
when gNSI is enabled {
  deviation path {
    deviate not-supported;
  }
}
```

But, if only `when` is allowed to be used, then the only thing we can do is to guard the AAA leaves themselves. However, this doesn't work as-is because it introduces a cyclic-dependency of `openconfig-system` -> `openconfig-aaa` **-> (new dep)** `openconfig-system-grpc` -> `openconfig-system` after modification to the ssh-key definition, for example:
```yang
    leaf ssh-key {
      when "not(boolean(../../../../../../oc-sys-grpc:grpc-servers" +
          "/oc-sys-grpc:grpc-server/oc-sys-grpc:config" +
          "/oc-sys-grpc:services[contains(., 'oc-gnsi:GNSI')]))";
      type string;
      description
        "SSH public key for the user (RSA or DSA)";
    }
```

One workaround is to define these particular SSH keys in a separate module and augment into openconfig-system, OR make the openconfig-aaa module augment into openconfig-system instead of openconfig-system using `use` for these leaves.

The simplest solution would be just to remove the deviate statements like the WiFi models did, and only have comments stating when these are not available. The YANG model would be less descriptive, but it doesn't in principle affect correctness.

My preference is to delete the deviate statements if having them doesn't provide a strong value for the user.

Reply:
@earies and others, what do you think about removing the deviate statements and using only descriptions for this scenario vs. using `when xxx deviate not-supported`?   Seems like it might be better to add the `when` to the intended leafs in the `/system/aaa/authentication/users/user/` subtree?

I do like the idea that the `when` clause, (similar to what @earies  hinted earlier) should key off a new set of identities (such as `oc-grpc-server-types`) defined for the well known grpc services.  Then we can have a very easy to understand statement such as:

```
          container users {
            when "../grpc-servers/type = 'oc-grpc-server-types:GNSI'" {
```

Reply:
Please do not use `when ... deviate` -- this is outside of the style guide.

Reply:
Reading https://datatracker.ietf.org/doc/html/rfc6020#section-7.19.5 I see we don't need deviate.  We achieve our intent using the `when` statement without any deviation.  We just need what you suggested earlier @wenovus.  Replace each of the existing "when xxx deviate not-supported" nodes with its equivalent: A when statement applied to the node we want to make conditional:
```
    leaf ssh-key {
      when "not(boolean(../../../../../../oc-sys-grpc:grpc-servers" +
          "/oc-sys-grpc:grpc-server/oc-sys-grpc:config" +
          "/oc-sys-grpc:services[contains(., 'oc-gnsi:GNSI')]))";
      type string;
      description
        "SSH public key for the user (RSA or DSA)";
    }
```

However, I think we want this conditional only if GNSI is actually enabled.   Since there is an identity involved, I am not sure how to write the xpath expression.  Any tips @wenovus ?  
Intent:   "when /system/grpc-servers/grpc-server[GNSI]/state/enable = TRUE"

Reply:
One side effect that is undesirable, but perhaps unavoidable is adding when statements into openconfig-aaa which refer to gnsi requires openconfig-aaa to import the gnsi module.  It would be cleaner if there was some way to contain all "gnsi things" in the gnsi module.  But I don't see a way to do this?

Reply:
Given that `when ... deviate` should not be used, I'm going to recommend avoiding using any deviations because I anticipate we cannot avoid violating RFC7951 backward-incompatibility:

Firstly, the xpath expression is not an issue, I believe we can do this via
```yang
    leaf ssh-key {
      when "../../../../../../oc-sys-grpc:grpc-servers" +
          "/oc-sys-grpc:grpc-server/oc-sys-grpc:config" +
          "[contains(services, 'oc-gnsi:GNSI')]/enable = 'true'";
      type string;
      description
        "SSH public key for the user (RSA or DSA)";
    }
```

> One side effect that is undesirable, but perhaps unavoidable is adding when statements into openconfig-aaa which refer to gnsi requires openconfig-aaa to import the gnsi module. It would be cleaner if there was some way to contain all "gnsi things" in the gnsi module. But I don't see a way to do this?

However, for the import cycle issue, the only options to proceed if we want the deviations is to either,
1) change openconfig-aaa to `augment` into openconfig-system instead of being imported and instantiated via `uses` statements by openconfig-system.
2) change openconfig-system-grpc to be instantiated by openconfig-system via `uses` statements rather than `augment`.

This is because refactoring the gNSI module doesn't address the import cycle.

To illustrate, there are two different import cycles created when the above `when` is used:

![gNSI when-statement import cycle situation](https://github.com/openconfig/public/assets/50884368/4a6a0d47-69d0-4eee-98ec-5cba597f11bf)

It doesn't matter how we refactor gNSI -- it is the existing imports that are causing the issue.

However, neither option above is palatable due to the following impacts:
1) It would change the namespace of openconfig-aaa nodes from openconfig-system to openconfig-aaa. This affects the JSON output of RFC7951 marshal operations.
2) It would force all existing openconfig-system implementers to also implement openconfig-system-grpc module. It would also change the namespace of openconfig-system-grpc nodes from openconfig-system-grpc to openconfig-system, which affects the JSON output of RFC7951 marshal operations.

Though ygot tooling ignores RFC7951 namespaces during unmarshalling, there is no guarantee that existing devices may not reject configuration that supplies an incorrect namespace for YANG nodes.

---------

Given the above evidence I will remove the deviations, and claim that they're not practical to keep.

Reply:
I agree that removing the deviations and updating the credentialz module description to declare the relevant paths are invalid is good enough for now.  A future PR can explore how to more explicitly model this mutual exclusion.